skfp/applicability_domain/knn.py

```diff
@@ -0,0 +1,242 @@
+    >>> import numpy as np
+    >>> X_train_binary = np.array([
+    ...     [1, 1, 1],
+    ...     [0, 1, 1],
+    ...     [0, 0, 1]
+    ... ])
+    >>> X_test_binary = 1 - X_train_binary
+    >>> knn_ad_checker_binary = KNNADChecker(k=2, metric="tanimoto_binary_distance", agg="mean")
+    >>> knn_ad_checker_binary
+    KNNADChecker()
+
+    >>> knn_ad_checker_binary.fit(X_train_binary)
+    KNNADChecker()
+
+    >>> knn_ad_checker_binary.predict(X_test_binary)
+    array([False, False, False])
+
+    >>> X_train_count = np.array([
+    ...     [1.2, 2.3],
+    ...     [3.4, 4.5],
+    ...     [5.6, 6.7]
+    ... ])
+    >>> X_test_count = X_train_count + 10
+    >>> knn_ad_checker_count = KNNADChecker(k=2, metric="tanimoto_count_distance", agg="min")
+    >>> knn_ad_checker_count
+    KNNADChecker()
+
+    >>> knn_ad_checker_count.fit(X_train_count)
+    KNNADChecker()
+
+    >>> knn_ad_checker_count.predict(X_test_count)
+    array([False, False, False])
+
+    """
+
+    _parameter_constraints: dict = {
+        **BaseADChecker._parameter_constraints,
+        "k": [Interval(Integral, 1, None, closed="left")],
+        "metric": [
+            callable,
+            StrOptions(METRIC_NAMES),
+        ],
+        "threshold": [None, Interval(Real, 0, 1, closed="both")],
+    }
+
+    def __init__(
+        self,
+        k: int,
+        metric: str | Callable = "tanimoto_binary_distance",
+        agg: str = "mean",
+        threshold: float = 0.95,
+        n_jobs: int | None = None,
+        verbose: int | dict = 0,
+    ):
+        super().__init__(
+            n_jobs=n_jobs,
+            verbose=verbose,
+        )
+        self.k = k
+        self.metric = metric
+        self.agg = agg
+        self.threshold = threshold
+
+    def _validate_params(self) -> None:
+        super()._validate_params()
+        if isinstance(self.metric, str) and self.metric not in METRIC_FUNCTIONS:
+            raise InvalidParameterError(
+                f"Allowed metrics: {METRIC_NAMES}. Got: {self.metric}"
+            )
+        if isinstance(self.agg, str) and self.agg not in ["mean", "max", "min"]:
+            raise InvalidParameterError("Unknown aggregration method.")
+
+    def fit(  # noqa: D102
+        self,
+        X: np.ndarray,
+        y: np.ndarray | None = None,  # noqa: ARG002
+    ):
+        X = validate_data(self, X=X)
+        if self.k > X.shape[0]:
+            raise ValueError(
+                f"k ({self.k}) must be smaller than or equal to the number of training samples ({X.shape[0]})"
+            )
+
+        self.X_train_ = X
+        self.k_used = 1 if self.agg == "min" else self.k
+
+        if isinstance(self.metric, str) and self.metric in SKFP_BULK_METRIC_NAMES:
+            bulk_func = SKFP_BULK_METRICS[self.metric]
+            dist_mat = bulk_func(X, X)
+            np.fill_diagonal(dist_mat, np.inf)
+            k_nearest = np.partition(dist_mat, self.k_used, axis=1)[:, : self.k_used]
```

Comment:
I think we can assume that only vector functions are allowed, not bulk ones. This will simplify the code, and should be fast enough